stratum/hal/lib/tdi/tdi_lut_manager.cc

```diff
@@ -0,0 +1,208 @@
+      auto mk = *it;
+      switch (mk.field_match_type_case()) {
+        case ::p4::v1::FieldMatch::kExact: {
+          CHECK_RETURN_IF_FALSE(!IsDontCareMatch(mk.exact()));
+          RETURN_IF_ERROR(
+              table_key->SetExact(mk.field_id(), mk.exact().value()));
+          break;
+        }
+        case ::p4::v1::FieldMatch::kTernary: {
+          CHECK_RETURN_IF_FALSE(!IsDontCareMatch(mk.ternary()));
+          RETURN_IF_ERROR(table_key->SetTernary(
+              mk.field_id(), mk.ternary().value(), mk.ternary().mask()));
+          break;
+        }
+        default:
+          RETURN_ERROR(ERR_INVALID_PARAM)
+              << "Invalid or unsupported match key: " << mk.ShortDebugString();
+      }
+    }
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status TdiLutManager::DoWriteVlutEntry(
+    std::shared_ptr<TdiSdeInterface::SessionInterface> session,
+    uint32 bfrt_table_id, const ::p4::v1::Update::Type type,
+    const ::p4::v1::MvlutEntry& mvlut_entry) {
+  // Lock is already acquired by the caller
+  CHECK_RETURN_IF_FALSE(type != ::p4::v1::Update::UNSPECIFIED)
+      << "Invalid update type " << type;
+
+  ASSIGN_OR_RETURN(auto table_key,
+                   tdi_sde_interface_->CreateTableKey(bfrt_table_id));
+  RETURN_IF_ERROR(BuildTableKey(mvlut_entry, table_key.get()));
+
+  ASSIGN_OR_RETURN(
+      auto table_data,
+      tdi_sde_interface_->CreateTableData(bfrt_table_id, 0));
+
+  if (type == ::p4::v1::Update::INSERT || type == ::p4::v1::Update::MODIFY) {
+    for (const auto& param : mvlut_entry.param().params()) {
+      RETURN_IF_ERROR(table_data->SetParam(param.param_id(), param.value()));
+    }
+  }
+
+  switch (type) {
+    case ::p4::v1::Update::INSERT: {
+      RETURN_IF_ERROR(tdi_sde_interface_->InsertTableEntry(
+          device_, session, bfrt_table_id, table_key.get(),
+	  table_data.get()));
+      break;
+    }
+    case ::p4::v1::Update::MODIFY: {
+      RETURN_IF_ERROR(tdi_sde_interface_->ModifyTableEntry(
+          device_, session, bfrt_table_id, table_key.get(),
+	  table_data.get()));
+      break;
+    }
+    case ::p4::v1::Update::DELETE: {
+      RETURN_IF_ERROR(tdi_sde_interface_->DeleteTableEntry(
+          device_, session, bfrt_table_id, table_key.get()));
+      break;
+    }
+    default:
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported update type: " << type;
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status TdiLutManager::ReadTableEntry(
+    std::shared_ptr<TdiSdeInterface::SessionInterface> session,
+    const ::p4::v1::ExternEntry& entry,
+    WriterInterface<::p4::v1::ReadResponse>* writer) {
+  absl::ReaderMutexLock l(&lock_);
+
+  if (entry.extern_type_id() == kMvlutExactMatch ||
+      entry.extern_type_id() == kMvlutTernaryMatch) {
+      ::p4::v1::MvlutEntry mvlut_entry;
+      CHECK_RETURN_IF_FALSE(entry.entry().UnpackTo(&mvlut_entry))
+          << "Entry " << entry.ShortDebugString()
+          << " is not a mvlut entr.";
```

Comment:
`entr`?

Reply:
Also, `mvlut` would take the article `an`, not `a`, because one would pronounce `mvlut` EM-VEE-LUT, giving it an initial vowel sound.